dynamic_config/dynamic_config_manager.py

```diff
@@ -0,0 +1,196 @@
+  """
+  cluster_template = json_format.Parse(
+      json.dumps(yaml.load(cluster_template_string, Loader=yaml.FullLoader)), cluster_pb2.Cluster())
+
+  mutate_actions = list(Action)
+
+  def __init__(self, config: cluster_manager_pb2.DynamicClusterConfigManagerSettings):
+    """Create a DynamicClusterConfigManager."""
+    DynamicConfigManager.__init__(self)
+    self.active_config = discovery_pb2.DiscoveryResponse()
+    self.active_config.version_info = "1"
+    self.inactive_clusters = []
+    clusters = []
+    for service_config in config.clusters:
+      new_cluster = cluster_pb2.Cluster()
+      new_cluster.CopyFrom(self.cluster_template)
+      new_cluster.load_assignment.cluster_name = new_cluster.name = service_config.name
+      for endpoint_config in service_config.endpoints:
+        host = new_cluster.load_assignment.endpoints[0].lb_endpoints.add()
+        socket_address = host.endpoint.address.socket_address
+        socket_address.address = endpoint_config.ip
+        socket_address.port_value = int(endpoint_config.port)
+      clusters.append(new_cluster)
+
+    # Randomly assign clusters to either inactive or active.
+    while clusters:
+      add_as_active_cluster = random.choice([True, False])
+      if add_as_active_cluster:
+        self.active_config.resources.add().Pack(clusters.pop())
+      else:
+        self.inactive_clusters.append(clusters.pop())
+    self.output_file = config.output_file if config.output_file else "new_cds.pb"
+    # TODO(kbaichoo): Handle nanos get ignored
+    self.refresh_interval = config.refresh_interval.ToSeconds()
+    # Tracks the last mutated action. Used for testing.
+    self.last_mutate_action = None
+
+  def mutate(self):
+    """Invoke to randomly mutate the current cluster configuration."""
+    action = random.choice(self.mutate_actions)
+    self.last_mutate_action = action
+    if action == self.__class__.Action.ADD:
+      # Randomly pick a number of clusters to add
+      num_clusters_to_add = random.randrange(0, len(self.inactive_clusters) + 1)
+
+      # All clusters active or 0 clusters choosen, do nothing.
+      if not self.inactive_clusters or not num_clusters_to_add:
+        return
+
+      random.shuffle(self.inactive_clusters)
+
+      for i in range(num_clusters_to_add):
+        self.active_config.resources.add().Pack(self.inactive_clusters.pop())
+
+    elif action == self.__class__.Action.REMOVE:
+      resources = self.active_config.resources
+      num_clusters_to_remove = random.randrange(0, len(resources) + 1)
+
+      # All clusters inactive or 0 clusters choosen, do nothing.
+      if not resources or not num_clusters_to_remove:
+        return
+
+      for i in range(num_clusters_to_remove):
+        any_resource = resources.pop(random.randrange(0, len(resources)))
+        unpacked_cluster = cluster_pb2.Cluster()
+        any_resource.Unpack(unpacked_cluster)
+        self.inactive_clusters.append(unpacked_cluster)
```

Comment:
optional: extracting REMOVE and ADD into functions might make this more readable.

Reply:
I tend to agree with @tomjzzhang, we don't want to extract them for the purpose of reuse, but to make the high level business logic of mutate more readable. That way the reader doesn't have to delve into the details immediately.